### DISCUSSION TITLE
How to write a dependent test cases in Foundry

### DISCUSSION THREAD

**Discussion by devrajsinghrawat** [Cyfrin/foundry-full-course-cu#555] [Q&A]
Hi Friends,

I am trying to **test a scenario where we need might need to execute test case with the context of previous test cases**.

If I am correct then in case  of JavaScript we can achieve it by writing multiple **`it blocks`** within single **`describe block`** (example show below)

```
describe('test dependent test cases', () => {
  it('deposit by Alice', () => {
    // Write assertions here
    
  });

  it('deposit by Bob', () => {
    // Write assertions here
    
  });

  it('fetch all depositors', () => {
    // Write assertions here
    
  });
});
```

I would like to understand how can we **`achieve the same in case of foundry`**. I had written an example by calling the previous test case in next test case but not sure if that the correct approach.

**Pls Ignore the Test case validation part,** and focus on the idea to understand how better we can write a dependent test cases

![Screenshot 2023-08-06 at 1 08 46 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/df7a767c-cb80-488f-b34a-91f1fc8174af)

Contract FundMe
![Screenshot 2023-08-06 at 1 18 36 PM](https://github.com/Cyfrin/foundry-full-course-f23/assets/44597990/542d6f8f-614d-40ae-a16d-1e6008e1455a)

**Comment by shikhar229169** [comment]
There is a concept in foundry called stateful fuzz testing and its tutorial is in lesson 12.
It solves the same problem that you want to solve, which is the current test should depend on the previous contract state which was a result of a particular test.

So, I suggest you to currently follow along with the course and when you learn that concept in lesson 12, again come back to all previous lessons and apply that.

**Comment by theirrationalone** [comment]
If i'm right, You're talking about that "beforeEach("do something..", asyn() =>{after_work... }) ;" stuff, to set some previous states/setting some requirements before testing any function....??? 

If do so... 

Then in foundry you can set all previous states at the end of "setUp" function Or also can use "modifiers" Before each function. Or... 

If want to fire random data in random order in functions expected to test,

then "stateful fuzz testing Or invariant fuzz testing (lesson 12: invariant fuzz testing)", would be a great choice. 

Keep it in mind: data will be populated in random order, It's not necessary that the data populated in previous test would also be populated and persisted for next test in respect of previous test. 


If you want to find vulnerabilities in your contracts then invariant fuzz testing is Great, Therefore It can touch the edges. 

Or if you want to be persistent with your data then you should use modifiers Or setUp function.
